lib/trino-parquet/src/main/java/io/trino/parquet/reader/ParquetReader.java

```diff
@@ -306,7 +306,9 @@ private boolean advanceToNextRowGroup()
             long rowCount = currentGroupRowRanges.getRowCount();
             columnIndexRowsFiltered += currentGroupRowCount - rowCount;
             if (rowCount == 0) {
-                return false;
+                // Filters on multiple columns with page indexes may yield non-overlapping row ranges and eliminate the entire row group.
+                // Advance to next row group to ensure that we don't return a null Page and close the page source before all row groups are processed
+                return advanceToNextRowGroup();
```

Comment:
We should probably rename the method to `advanceToNextMatchingRowGroup`
`advanceToNextRowGroup` is rather misleading because it naively implies that only one advancement is made, and not potentially more.

Reply:
These not empty row groups, they're row groups which have been pruned due to non-overlapping row ranges produced by filters on multiple columns.
We also don't check row count/emptiness when there isn't a column index. We don't expect to see an empty row group as there's no reason for the writer to produce that that but it's not technically impossible.
I don't want to complicate the method name too much by specifying the precise conditions in front of "Next".